### PR TITLE
Submit education benefits form

### LINKED ISSUE #2995: Submit education benefits form
Submit education benefits form filled out by a veteran to the api for processing:

- [x] Convert form from form field format to json schema
- [x] Make fetch (or whatever we use to make api calls) call to api

### PR DESCRIPTION
**Pull request by jbalboni**
Closes #2995.

This gets us to a mostly working form submit action. "Mostly" because there are still some schema changes that need to happen on the backend.

It's bigger than I expected it to be and does not include json schema validation in tests yet.

The changes in here are:
- Convert form data into valid application data (veteranToApplication in veteran.js). This is the core of these changes
- Form page and field updates to match schema. We weren't matching the schema in places or were missing required fields.
- Update validation. We were missing validation for some pages and fields, so that's been added.
- Other bug fixes that were noticed in the submission process.

### CODE REVIEW

File: src/js/common/components/form-elements/GrowableTable.jsx [hunk 1]
```diff
@@ -99,11 +100,8 @@ class GrowableTable extends React.Component {
                   { data: obj,
                     view: 'collapsed',
                     onValueChange: (subfield, update) => {
-                      const rows = this.props.rows.slice();
-                      const newRow = Object.assign({}, rows[index]);
-                      newRow[subfield] = update;
-                      rows[index] = newRow;
-                      this.props.onRowsUpdate(rows);
+                      const newRows = set(`[${index}].${subfield}`, update, this.props.rows);
```
Comment (jbalboni) [inline src/js/common/components/form-elements/GrowableTable.jsx]: Did this so we could update nested fields in table rows.

File: src/js/common/components/NavButtons.jsx [hunk 2]
```diff
@@ -21,7 +21,7 @@ export default class NavButtons extends React.Component {
     }
   }
   handleSubmit() {
-    if (this.props.isValid) {
+    if (this.props.canSubmit) {
```
Comment (jbalboni) [inline src/js/common/components/NavButtons.jsx]: isValid is for specific pages, needed something for the whole form.

File: src/js/edu-benefits/components/education-history/EducationHistoryFields.jsx [hunk 3]
```diff
@@ -35,29 +35,29 @@ export default class EducationHistoryFields extends React.Component {
             year={year}
             onValueChange={(update) => {this.props.onStateChange('highSchoolOrGedCompletionDate', update);}}/>
       </div>
-      {isValidDate(day.value, month.value, year.value)
```
Comment (jbalboni) [inline src/js/edu-benefits/components/education-history/EducationHistoryFields.jsx]: When I looked at the wireframes, the table wasn't conditional. I can remove this change if it is actually supposed to be conditional.

File: src/js/edu-benefits/utils/validations.js [hunk 4]
```diff
@@ -284,8 +288,33 @@ function isValidDirectDepositPage(data) {
...
+  return isValidField(isValidMonetaryValue, data.amount) && isValidField(isValidYear, data.year);
+}
+
+function isValidRotcHistoryPage(data) {
+  return data.seniorRotc.rotcScholarshipAmounts.every(isValidRotcScholarshipAmount);
+}
+
 function isValidForm(data) {
-  return isValidBenefitsInformationPage(data);
+  return isValidBenefitsInformationPage(data)
```
Comment (alexose) [inline src/js/edu-benefits/utils/validations.js]: I'm wary of the amount of repetition happening in validations.js-- Given that each field has validation rules attached to it, It seems like we could iterate through all of them?

On the other hand, this does make it easier to introduce new higher-level validation rules, which I could see us doing...

File: src/js/edu-benefits/utils/veteran.js [hunk 5]
```diff
@@ -178,3 +184,130 @@ export function createVeteran() {
     }
   };
 }
+
+function convertFieldToValue(field) {
```
Comment (alexose) [inline src/js/edu-benefits/utils/veteran.js]: I'd like us not to diverge too far from what we do for HCA: https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/js/common/model/veteran.js#L700-L768

File: src/js/edu-benefits/components/education-history/EducationHistoryFields.jsx (hunk 3 above)
Comment (alexose) [inline src/js/edu-benefits/components/education-history/EducationHistoryFields.jsx]: Huh, you're right.  I could have sworn it was.  Maybe I made that up...

File: src/js/edu-benefits/components/military-history/MilitaryServiceReview.jsx [hunk 6]
```diff
@@ -62,11 +66,11 @@ export default class MilitaryServiceReview extends React.Component {
                 : null}
               <tr>
                 <td>From date:</td>
-                <td>{tour.fromDate.month.value ? `${tour.fromDate.month.value}/${tour.fromDate.day.value}/${tour.fromDate.year.value}` : null}</td>
+                <td>{displayDateIfValid(tour.dateRange.from)}</td>
```
Comment (alexose) [inline src/js/edu-benefits/components/military-history/MilitaryServiceReview.jsx]: :D

File: src/js/edu-benefits/components/military-history/MilitaryServiceReview.jsx [hunk 7]
```diff
@@ -62,11 +66,11 @@ export default class MilitaryServiceReview extends React.Component {
...
               </tr>
               <tr>
                 <td>To date:</td>
-                <td>{tour.toDate.month.value ? `${tour.toDate.month.value}/${tour.toDate.day.value}/${tour.toDate.year.value}` : null}</td>
+                <td>{displayDateIfValid(tour.dateRange.to)}</td>
```
Comment (alexose) [inline src/js/edu-benefits/components/military-history/MilitaryServiceReview.jsx]: :D x2

**Review by alexose — CHANGES_REQUESTED** [review]
This is looking great so far.  I would definitely focus on the veteranToApplication / convertFieldToValue thing, though, since it has the potential to be difficult to test.

Another thing I'd consider thinking about is validating against the actual JSONSchema on the client.  It might not be too hard, and it could save us a lot of headaches down the road.

**Review by alexose — APPROVED** [review]

File: src/js/edu-benefits/utils/validations.js (hunk 4 above)
Comment (jbalboni) [inline src/js/edu-benefits/utils/validations.js]: You mean iterating through the React fields? I think that would be tough to get working.

Or something else?

File: src/js/edu-benefits/utils/veteran.js (hunk 5 above)
Comment (jbalboni) [inline src/js/edu-benefits/utils/veteran.js]: I updated this to match the hca approach.

I really dislike Javascript's case statements, but it is simpler.

File: src/js/edu-benefits/utils/validations.js (hunk 4 above)
Comment (alexose) [inline src/js/edu-benefits/utils/validations.js]: Not necessarily going through the actual fields, but maybe keeping a dictionary of `field: validation` and looping through them.

Probably more trouble than it's worth--  Plus we might be able to rework a bunch of this when we validate against the schema directly.

**Comment by alexose** [comment]
Looks good.  I'll merge this in after CI

**Pull request merged**